transaction.js

```diff
@@ -0,0 +1,148 @@
+    // See http://bit.ly/Node_startSession for the startSession() docs
+    const session = client.startSession();
+
+    // Step 2: Optional. Define options to use for the transaction
+    const transactionOptions = {
+        readPreference: 'primary',
+        readConcern: { level: 'local' },
+        writeConcern: { w: 'majority' }
+    };
+
+    try {
+        // Step 3: Use withTransaction to start a transaction, execute the callback, and commit (or abort on error)
+        // Note: The callback for withTransaction MUST be async and/or return a Promise.
+        // See http://bit.ly/Node_withTransaction for the withTransaction() docs
+        const transactionResults = await session.withTransaction(async () => {
+
+            // Important:: You must pass the session to each of the operations   
+
+            // Add a reservation to the reservations array for the appropriate document in the users collection
+            const usersUpdateResults = await usersCollection.updateOne(
+                { email: userEmail },
+                { $addToSet: { reservations: reservation } },
+                { session });
+            console.log(`${usersUpdateResults.matchedCount} document(s) found in the users collection with the email address ${userEmail}.`)
+            console.log(`${usersUpdateResults.modifiedCount} document(s) was/were updated to include the reservation.`);
+
+            // Check if the Airbnb listing is already reserved for those dates. If so, abort the transaction.
+            const isListingReservedResults = await listingsAndReviewsCollection.findOne(
+                { name: nameOfListing, datesReserved: { $in: reservationDates } },
+                { session });
+            if (isListingReservedResults) {
+                await session.abortTransaction();
+                console.error("This listing is already reserved for at least one of the given dates. The reservation could not be created.");
+                console.error("Any operations that already occurred as part of this transaction will be rolled back.")
+                return;
+            }
+
+            //  Add the reservation dates to the datesReserved array for the appropriate document in the listingsAndRewiews collection
+            const listingsAndReviewsUpdateResults = await listingsAndReviewsCollection.updateOne(
+                { name: nameOfListing },
+                { $addToSet: { datesReserved: { $each: reservationDates } } },
+                { session });
+            console.log(`${listingsAndReviewsUpdateResults.matchedCount} document(s) found in the listingsAndReviews collection with the name ${nameOfListing}.`)
+            console.log(`${listingsAndReviewsUpdateResults.modifiedCount} document(s) was/were updated to include the reservation dates.`);
+
+        }, transactionOptions);
+
+        // TODO:  I couldn't find documentation that explicitly said that the transactionResults will be undefined if the transaction is 
```

Comment:
@mbroadst Can you confirm?

Reply:
`session.withTransaction` will return a Promise, which is rejected on error, and resolved on success. If `withTransaction` fails and you are using `async/await` then it will throw an error, so you'd have to wrap this call with a `try/catch` to get what you're looking for.

Reply:
I talked with Matt on Slack to clarify this, and I've updated the code.